axum/src/serve.rs

```diff
@@ -0,0 +1,132 @@
+use std::{
+    convert::Infallible,
+    io,
+    net::SocketAddr,
+    task::{Context, Poll},
+};
+
+use axum_core::{body::Body, extract::Request, response::Response};
+use futures_util::{future::poll_fn, FutureExt};
+use hyper1::server::conn::http1;
+use tokio::net::{TcpListener, TcpStream};
+use tower_hyper_http_body_compat::{HttpBody04ToHttpBody1, HttpBody1ToHttpBody04};
+use tower_service::Service;
+
+/// TODO(david): docs
+pub async fn serve<M, S>(tcp_listener: TcpListener, mut make_service: M) -> io::Result<()>
+where
+    M: for<'a> Service<&'a (TcpStream, SocketAddr), Error = Infallible, Response = S>,
```

Comment:
note to self: Probably should make a struct instead of `&'a (TcpStream, SocketAddr)`. It's not clear that the `SocketAddr` is `local_addr`.